anndata/_core/anndata.py

```diff
@@ -832,34 +488,27 @@ def _init_as_actual(
         self._check_dimensions()
         self._check_uniqueness()
 
-        # raw
-        if raw is None:
-            self._raw = None
-        elif isinstance(raw, Raw):
-            self._raw = raw
-        else:
-            # is dictionary from reading the file, nothing that is meant for a user
-            if self.isbacked:
-                raw_key = "raw/X" if "raw/X" in self.file else "raw.X"
-                if raw_key not in self.file:
-                    raise KeyError(
-                        f"Tried to check shape of raw from {self.file.filename}, but there was no entry for 'raw/X'."
-                    )
-                if isinstance(self.file[raw_key], h5py.Group):
-                    shape = self.file[raw_key].attrs["shape"]
-                else:
-                    shape = self.file[raw_key].shape
-            else:
-                shape = raw["X"].shape
-
-            self._raw = Raw(
-                self,
-                X=raw.get("X", None),
-                var=_gen_dataframe(
-                    raw['var'], shape[1], ['var_names', 'col_names']
-                ),
-                varm=raw['varm'] if 'varm' in raw else None,
-            )
+        if self.filename:
+            assert not isinstance(
+                raw, Raw
+            ), "got raw from other adata but also filename?"
+            from ..readwrite.h5ad import _read_raw, _read_legacy_raw_h5ad
+
+            # Read raw dict from file
+            raw, raw_override = {}, raw
+            if "raw" in self.file:
+                raw = _read_raw(self.file)
+            _read_legacy_raw_h5ad(self.file, raw, self.filename)
+
+            # override retrieved entries with specified ones
+            if raw_override:
+                raw.update(raw_override)
+        if raw and isinstance(raw, cabc.Mapping):
```

Comment:
I still think if you're checking if the value is `None` that should be explicit.

As an appeal to authority, the [google python style guide](http://google.github.io/styleguide/pyguide.html#2144-decision) and [pep8](https://www.python.org/dev/peps/pep-0008/#programming-recommendations) say this too.

Reply:
What you linked agrees with me:

> Google: Use the “implicit” false if possible, e.g., `if foo`: rather than `if foo != []`

> PEP8: Also, beware of writing `if x` when you really mean `if x is not None` – e.g. when testing whether a variable or argument that defaults to None was set to some other value.

I precisely mean “if raw is not empty”, elegantly expressed in Python as `if raw`: `raw` can either be `{something}` or `{}` or `None` and I want to treat the last two cases in the same way.

Using `if x` is like using `if x == y`: It (`__bool__`/`__eq__`) relies on user-specified behavior and it’s pythonic to use it. If you propose to not rely on `__bool__` on the grounds of numpy overriding it, we also need to get rid of our use of `__eq__` in our code base. I say we keep using both, because we usually know if something might be a numpy array or not.

Being specific about `is None` is useful for one purpose: If you want to treat the `None` case differently from each other empty case. That’s common if you have one of those (overlapping) use cases:

1. have `param=None` as default value and want some custom, smart behavior.
2. want to treat `None` differently from other falsy values

In this case, `raw={}` and `raw=None` mean exactly the same, so we use `if raw`.